src/Route.php

```diff
@@ -71,17 +72,13 @@ class Route
      * @param string $path Path to match.
      * @param string|callable|MiddlewareInterface $middleware Middleware to use when this route is matched.
      * @param int|array $methods Allowed HTTP methods; defaults to HTTP_METHOD_ANY.
-     * @param string|null $name the route name
+     * @param null|string $name the route name
      * @throws Exception\InvalidArgumentException for invalid path type.
      * @throws Exception\InvalidArgumentException for invalid middleware type.
      * @throws Exception\InvalidArgumentException for any invalid HTTP method names.
      */
-    public function __construct($path, $middleware, $methods = self::HTTP_METHOD_ANY, $name = null)
+    public function __construct(string $path, $middleware, $methods = self::HTTP_METHOD_ANY, string $name = null)
```

Comment:
Would be nice if we could typehint on `$middleware` and `$methods`.
I've seen somewhere inconsistent with `$methods`, I think in some place I've seen `=null` which was translated later on to `ANY`.

What with `$middleware`. Is it not now just `MiddlewareInterface`?

Reply:
I was just thinking about this myself when I reviewed your earlier patch, and just dove into how zend-expressive uses it.

At this point, within zend-expressive itself, we _always_ pass a middleware instance, as `Application::prepareMiddleware()` creates an instance based on the argument passed to `Application::route()`. As such, I think we could likely update this to have a typehint.

We cannot typehint `$methods` currently, as it can be either an array of HTTP method names **or** the `HTTP_METHOD_ANY` constant, which is an integer. I did it this way as the HTTP specification allows arbitrary method names, so there's no way to create a "master list" representing all methods.

Reply:
Updated to `MiddlewareInterface` typehint on `$middleware`.

About `$methods`, please see:
https://github.com/zendframework/zend-expressive/pull/530/files#diff-c9248b3167fc44af085b81db2e292837R283
So there we have `array $methods = null` and then we have:
```php
$methods = null === $methods ? Router\Route::HTTP_METHOD_ANY : $methods;
```

Maybe we should do the same?
I'm not sure if I like this behaviour (`null` means any, array defines method explicitly, so emoty array means none).
Similar solution we have on mocking in PHPUnit - `setMethods()` - we can pass `null|array` and it has different behaviour with `null` and empty array.

I would prefer to have it consistent - the same in expressive and expressive-router.

Reply:
Nullable would work here, I guess; we'd have to change the value of the constant to be `null`, however, to ensure there's no breakage for folks who use the constant to specify the value. That would allow `?array $methods = null` in the signature.